syntax = "proto3";

package paxos;


service Acceptor {
    rpc Prepare(PrepareRequest) returns (PrepareResponse);
    rpc Accept(AcceptRequest) returns (AcceptResponse);
}

service Proposer {
    rpc Prepare(PrepareRequest) returns (PrepareResponse);
    rpc Accept(AcceptRequest) returns (AcceptResponse);
}

message PrepareRequest {
    int64 proposal_number = 1;
}

message PrepareResponse {
    bool promise = 1; 
    int64 proposal_number = 2;
    string accepted_value = 3; 
    int64 accepted_proposal_number = 4; 
}

message AcceptRequest {
    int64 proposal_number = 1;
    string value = 2;  
}

message AcceptResponse {
    bool accepted = 1; 
}

message ProposeRequest {
    string value = 1;
}

message ProposeResponse {
    bool success = 1;
}
